plugins/destination/mssql/client/migrate.go

```diff
@@ -2,48 +2,190 @@ package client
+		}
+	}
+	return result, tableChanges
+}
+
+func (*Client) canAutoMigrate(changes []schema.TableColumnChange) bool {
+	for _, change := range changes {
+		if change.Type == schema.TableColumnChangeTypeAdd && (change.Current.CreationOptions.PrimaryKey || change.Current.CreationOptions.NotNull) {
+			return false
+		}
+
+		if change.Type == schema.TableColumnChangeTypeRemove && (change.Previous.CreationOptions.PrimaryKey || change.Previous.CreationOptions.NotNull) {
+			return false
+		}
+
+		if change.Type == schema.TableColumnChangeTypeUpdate {
+			return false
+		}
+	}
+	return true
+}
+
+func (c *Client) execStatements(ctx context.Context, table *schema.Table, statements []string) error {
+	if len(statements) == 0 {
 		return nil
 	}
 
-	switch tableExist, err := c.tableExists(ctx, table); {
-	case err != nil:
-		return fmt.Errorf("failed to check if table %s exists: %w", table.Name, err)
-	case tableExist:
-		if err := c.autoMigrateTable(ctx, table); err != nil {
-			return err
+	return c.doInTx(ctx, func(tx *sql.Tx) error {
+		query := strings.Join(statements, "\n")
+		c.logger.Debug().Str("table", table.Name).Str("query", query).Msg("exec migration statement")
+		_, err := c.db.ExecContext(ctx, query)
+		return err
+	})
+}
+
+func (c *Client) autoMigrateTable(ctx context.Context, table *schema.Table, changes []schema.TableColumnChange) error {
+	var statements []string
+	for _, change := range changes {
+		if change.Type == schema.TableColumnChangeTypeAdd {
+			def := queries.GetDefinition(&change.Current, c.pkEnabled())
+			statements = append(statements, queries.AddColumn(c.schemaName, table, def))
 		}
-	default:
-		if err := c.createTable(ctx, table); err != nil {
-			return err
+	}
+
+	err := c.execStatements(ctx, table, statements)
+	if err != nil {
+		return err
+	}
+	return c.ensureTVP(ctx, table)
+}
+
+// Migrate relies on the CLI/client to lock before running migration.
+func (c *Client) Migrate(ctx context.Context, tables schema.Tables) error {
+	schemaTables, err := c.schemaTables(ctx, tables)
+	if err != nil {
+		return err
+	}
+
+	normalizedTables := c.normalizedTables(tables)
+
+	if c.spec.MigrateMode != specs.MigrateModeForced {
+		nonAutoMigrableTables, changes := c.nonAutoMigrableTables(normalizedTables, schemaTables)
+		if len(nonAutoMigrableTables) > 0 {
+			return fmt.Errorf("tables %s with changes %v require force migration. use 'migrate_mode: forced'", strings.Join(nonAutoMigrableTables, ","), changes)
+		}
+	}
+
+	for _, table := range normalizedTables {
+		c.logger.Info().Str("table", table.Name).Msg("Migrating table")
+		if len(table.Columns) == 0 {
+			c.logger.Info().Str("table", table.Name).Msg("Table with no columns, skipping")
+			continue
+		}
+		schemaTable := schemaTables.Get(table.Name)
+		if schemaTable == nil {
+			c.logger.Debug().Str("table", table.Name).Msg("Table doesn't exist, creating")
+			if err := c.createTable(ctx, table); err != nil {
+				return err
+			}
+		} else {
```

Comment:
I'd prefer `continue` statement in the `if` above, so that we don't have a lot of ident for `else`

Reply:
Done in https://github.com/cloudquery/cloudquery/pull/8318/commits/978a2215715f69c7766f31586e15262cbf08deeb